azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -0,0 +1,96 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This is a placeholder policy engine class to test that the regorus
+# dependency is correctly installed.
+# pylint: disable=too-few-public-methods
+
+import pwd
+import os
+import sys
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.version import get_distro
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.event import WALAEventOperation, add_event
+from azurelinuxagent.common import conf
+
+def log_policy_info(formatted_string, op=WALAEventOperation.Policy, send_event=True):
+    """Log information to console and telemetry."""
+    logger.info("[Policy] " + formatted_string)
+    if send_event:
+        add_event(op=op, message=formatted_string)
+
+
+class PolicyEngine:
+    """Base class for policy engine"""
+    _instance = None
+
+    @classmethod
+    def get_instance(cls):
+        if cls._instance is None:
+            cls._instance = cls()
+        return cls._instance
+    
+    @staticmethod
+    def _is_policy_supported():
+        distro_info = get_distro()
+        distro_name = distro_info[0]
+        try:
+            distro_version = FlexibleVersion(distro_info[1])
+        except ValueError:
+            return False
+        return distro_name.lower() == 'ubuntu' and distro_version.major >= 16
+
+    def __init__(self):
+        self._initialized = False
+        self._policy_supported = False
+        self._extension_policy_enabled = False
+        self._engine = None
+
+    def initialize(self):
+        try:
+            if self._initialized:
+                return
+            self._policy_supported = self._is_policy_supported()
+
+            if not self._policy_supported:
+                log_policy_info("Policy enforcement is not supported on {0}".format(get_distro()))
+                return
+
+            if not conf.get_extension_policy_enabled():
+                log_policy_info("Extension policy enforcement is disabled")
+                return
+
+            # TO DO - import code is for e2e testing, remove once the binary has been published to GA package
+            user = next((u.pw_name for u in pwd.getpwall() if u.pw_uid == 1000), None)
```

Comment:
This logic to add the executable to sys.path should probably be moved to the e2e test itself

Reply:
From what I can tell, the e2e tests and the agent have two different environments/path variables, maybe because the agent runs as root user and the test setup does not. Appending to PATH/PYTHONPATH in install-tools does not translate over to the agent's env, which causes import failures. I hardcoded the user ID here because "~" points to "/root" instead of "/home/<user>". Can you think of any workarounds?

Reply:
if you copy the binary to the "bin" folder instead of a new "executables" folder, the PATH should already be taken care of.

the SshClient runs set-agent-env, which adds "bin" to PATH

```
echo "Creating ~/bin/set-agent-env to set PATH and PYTHONPATH"

echo "
if [[ \$PATH != *\"$HOME/bin\"* ]]; then
  PATH=\"$HOME/bin:\$PATH:\"
fi
```

Reply:
actually, this is for the *.so library for regorus, right? should go to "lib", not "bin" as I suggested above,

if that is the case, then where will Python look for the *.so file? PATH or LD_LIBRARY_PATH? If the former, add "lib" to PATH in set-agent-env; if the latter, add it to LD_LIBRARY_PATH and add this variable to set-agent-env

Reply:
btw, regardless of how this is done, this should be done in the test code, not in the agent's code

Reply:
ah, i see... you did it in the agent because it's the agent's process who needs to load the library, not the tests.

still, since we are doing this for the tests only, it should be done on the test code for now. luckily, we are enabling this only on ubuntu, which uses systemd, so you can set the environment variables for the agent service according to your needs. the test can create a drop-in file to set the proper environment. if you are not familiar with drop-ins, @maddieford  can help you

Reply:
alternatively, you could decide what will be the path for the prod release and set its value here, then have the test setup move the binary that location.